### PR TITLE
Add GraalVM Native Image compilation support

### PR DESCRIPTION
**Pull request by zoewangg**
## Description
- Add GraalVM native image configurations to provide out-of-box support for native image compilation.
- Add `archetype-app-quickstart` maven archetype to make it easier to bootstrap client applications with v2 dependencies.
- Add `sdk-native-image-test` module for regression tests

## Motivation and Context
#1852, #1204 

## Testing
Tested locally by running the native image complied with sample applications using java sdk 2.x
~~I will add regression tests in this repo.~~
Added tests

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the **CONTRIBUTING** document
- [ ] Local run of `mvn install` succeeds
- [ ] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [ ] I have read the **README** document
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed
- [x] A short description of the change has been added to the **CHANGELOG**
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: archetypes/archetype-tools/src/test/resources/serviceMapping.vm [hunk 1]
```diff
@@ -0,0 +1 @@
..., "cloudhsm": "CloudHsm", "honeycode": "Honeycode", "cloudsearch": "CloudSearch", "acm": "Acm", "timestreamwrite": "TimestreamWrite", "inspector": "Inspector", "lakeformation": "LakeFormation", "elasticinference": "ElasticInference", "schemas": "Schemas", "amplify": "Amplify", "globalaccelerator": "GlobalAccelerator", "s3": "S3", "clouddirectory": "CloudDirectory", "accessanalyzer": "AccessAnalyzer", "greengrass": "Greengrass", "worklink": "WorkLink", "sagemaker": "SageMaker", "config": "Config", "codeartifact": "Codeartifact", "workspaces": "WorkSpaces", "savingsplans": "Savingsplans", "sqs": "Sqs", "iot": "Iot", "codestarconnections": "CodeStarConnections", "ssoadmin": "SsoAdmin", "sfn": "Sfn", "directory": "Directory", "marketplacecommerceanalytics": "MarketplaceCommerceAnalytics", "sns": "Sns", "mediastoredata": "MediaStoreData", "workmail": "WorkMail", "cognitoidentityprovider": "CognitoIdentityProvider", "codedeploy": "CodeDeploy", "iotjobsdataplane": "IotJobsDataPlane", "mobile": "Mobile", "docdb": "DocDb", "neptune": "Neptune", "wafv2": "Wafv2", "glacier": "Glacier", "apigatewaymanagementapi": "ApiGatewayManagementApi", "fsx": "FSx", "sagemakerruntime": "SageMakerRuntime"})
```
Comment (cenedhryn) [inline archetypes/archetype-tools/src/test/resources/serviceMapping.vm:1]: How is this file updated with new services?

File: test/sdk-native-image-test/README.md [hunk 2]
```diff
@@ -0,0 +1,18 @@
+# AWS SDK Native Image Test
+
+## Description
+This module contains a sample application using AWS SDK for Java 2.x, and it
+ used to test GraalVM Native Image compilation.
```
Comment (cenedhryn) [inline test/sdk-native-image-test/README.md:5]: NIT: should be "it `is` used"

File: buildspecs/archetype-native-image-test.yml [hunk 3]
```diff
@@ -0,0 +1,62 @@
+version: 0.2
```
Comment (cenedhryn) [inline buildspecs/archetype-native-image-test.yml:1]: Does anything happen when you run this buildspec, or do you execute target just to see that the app will start up?

**Review by cenedhryn — COMMENTED** [review]
The README.md for the `archetypes` folder should also add `archetype-app` to the list of archetypes.

File: archetypes/archetype-app/README.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+## Usage
+
+You can use `mvn archetype:generate` to generate a project using this archetype. See [maven archetype usage guidance][maven-archetype-usage] for more information.
+
+- Interactive mode
+
+```
+mvn archetype:generate \
+  -DarchetypeGroupId=software.amazon.awssdk \
+  -DarchetypeArtifactId=archetype-app \
```
[the remark this answers is not in the document]
Comment (zoewangg) [reply archetypes/archetype-app/README.md:17]: I knew it! 🔥 Any suggestions?

File: archetypes/archetype-app/README.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    -DartifactId=sample-project \
+    -Dservice=s3  \
+    -DinteractiveMode=false
+```
+
+### Parameters
+      
+Parameter Name | Default Value | Description
+---|---|---
+`service` (required) | n/a | Specifies the service client to be used in the lambda function, eg: s3, dynamodb. You can find available services [here][java-sdk-v2-services].
```
[the remark this answers is not in the document]
Comment (zoewangg) [reply archetypes/archetype-app/README.md:39]: Nope, only one service client is supported for now. I can update the description to be more clear

File: archetypes/archetype-app/src/main/resources/archetype-resources/README.md [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+#[[##]]# Development
+
+Below is the structure of the generated project.
+
+```
+├── src
+│   ├── main
+│   │   ├── java
+│   │   │   └── pacakge
```
[the remark this answers is not in the document]
Comment (zoewangg) [reply archetypes/archetype-app/src/main/resources/archetype-resources/README.md:18]: 🤦🏽 good catch

File: archetypes/archetype-tools/src/test/resources/serviceMapping.vm (hunk 1 above)
Comment (zoewangg) [reply archetypes/archetype-tools/src/test/resources/serviceMapping.vm:1]: This is just used for testing. serviceMapping.vm is created at build time

File: buildspecs/archetype-native-image-test.yml (hunk 3 above)
Comment (zoewangg) [reply buildspecs/archetype-native-image-test.yml:1]: Basically this verifies that archetype works, the generated application can be compiled to a native image and the native image can be executed

File: archetypes/archetype-app/README.md [hunk 7]
```diff
@@ -0,0 +1,53 @@
+# Maven Archetype for client applications using the AWS SDK for Java 2.x
+
+## Description
+This is an Apache Maven Archetype to create a client application template using
+ [AWS Java SDK 2.x][aws-java-sdk-v2]. You can use the archetype to generate
+  an application that has out-of-box [GraalVM Native Image][graalvm] support
```
[the remark this answers is not in the document]
Comment (zoewangg) [reply archetypes/archetype-app/README.md:6]: Yeah, I can add more description

> This is supposed to be our general entrypoint into generating a v2 application?!

Yeah, that's what I envisioned it.

File: archetypes/archetype-app/README.md [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+```
+
+### Parameters
+      
+Parameter Name | Default Value | Description
+---|---|---
+`service` (required) | n/a | Specifies the service client to be used in the lambda function, eg: s3, dynamodb. You can find available services [here][java-sdk-v2-services].
+`groupId`(required) | n/a | Specifies the group ID of the project
+`artifactId`(required) | n/a | Specifies the artifact ID of the project
+`nativeImage` | n/a | Specifies whether GraalVM Native Image configuration should be included
```
[the remark this answers is not in the document]
Comment (zoewangg) [reply archetypes/archetype-app/README.md:42]: From what I can tell, it's synonymous with GraalVM native image in Java land. Yeah, it's possible there might be another type of native image when project Leyden is finished, probably many, many years later. Other frameworks such as quarkus, Micronaut also use `native image`, so I think it should be fine.

File: archetypes/archetype-app/pom.xml [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+                            <goal>copy-resources</goal>
+                        </goals>
+                        <configuration>
+                            <outputDirectory>${basedir}/target/classes</outputDirectory>
+                            <encoding>UTF-8</encoding>
+                            <resources>
+                                <resource>
+                                    <directory>${basedir}/target/classes/archetype-resources</directory>
+                                    <includes>
+                                        <include>global.vm</include>
```
[the remark this answers is not in the document]
Comment (zoewangg) [reply archetypes/archetype-app/pom.xml:162]: Yeah, we need to copy the files to two different locations.

**Review by cenedhryn — APPROVED** [review]

**Pull request merged**